### ISSUE TITLE
[v2] .asyncTemplate() and Others template strategies

### ISSUE THREAD

**Issue by russiann** [framework7io/Framework7#1816]
This is a enhancement

Actually I need something like this:
```js
component: {
  asyncTemplate: (render) => {
    request(...)
       .then(html => {
         render(html)
       })
  }
}
```

There any way to exists this method?

Other idea is allow to return dom elements in `.render()` or/and `.template()` methods like this:

```js
component: {
  render: () => {
    return domElement
  },
  // or
  render: domElement
}
```
This would be awesome because allows developers to uses libs like `jsx-dom` or `vhtml` and do something beautiful like this:

```js
component: {
  render: () => {
    return (
      <div class="page">
        <div class="page-content">
          {items.map(item, =>
            <div class="whatever"></div>
          )}
        </div>
      </div>
    )
  },
}
```

**Comment by J05HI** [comment]
@russiann @nolimits4web
Definitely need that :)

**Comment by nolimits4web** [comment]
There is no much sense to have async templates because, there is no way to load component until it requested. There is an async routes for that covered in this article https://blog.framework7.io/mastering-v2-router-958ea2dbd24f. As for `render()` function it is already supported, you must return HTMLElement or string template inside of this function

**Comment by russiann** [comment]
@nolimits4web
Wow! I did not know about the `async` method. This definitely solves my problem. I will test it today. Thank you!!

**Comment by russiann** [comment]
@nolimits4web
I noticed that to use click events in the template I need to use an @ (`@click=`).

As you can see above I'm using JSX with vhtml.js and it only accept attributes with \`standard\` names. The @ causes an error :/ 

There any way to add a second way to do this? Like a `f7` prefix ( `f7-click=` )?

**Comment by nolimits4web** [comment]
@russiann not at the moment, i'll add `f-click` support

**Comment by nolimits4web** [comment]
@russiann does it support the following format `f7-on:click="doSomething"`?

**Comment by russiann** [comment]
@nolimits4web Not, the same error with `:`
Link to test: https://jsfiddle.net/developit/9q0vyskg/

**Comment by nolimits4web** [comment]
Shame, then it will `f7-on-...`

**Comment by nolimits4web** [comment]
well no, because some events have colon in names like `page:init` or `swipeout:open` so something like `f7-on-page:init` won't work as well

**Comment by russiann** [comment]
What about `f7-on="click:method"`?

**Comment by russiann** [comment]
`f7-on="page:init"`

**Comment by nolimits4web** [comment]
No, because it is support a bit more logic like @click="alert(5)". And same `f7-on="page:init:method"` looks bad and not really parseable

**Comment by russiann** [comment]
![sad](https://media.giphy.com/media/KDRv3QggAjyo/giphy.gif)

This is so sad...

**Comment by nolimits4web** [comment]
Agree, there probably could be some kind of babel jsx plugin to understand custom attributes. I see Vue js have their own babel jsx plugin.

But anyway, if you are going to use babel, you can just use es-next templates to achieve almost the same:
```js
...
render() {
  const self = this;
  const { items } = self;
  return `
    <div class="foo">
      <h1>Hi!</h1>
      <p @click="test">Here is a list of ${items.length} items:</p>
      <ul>
        ${items.map(item => `
          <li>${item}</li>
        `).join('')}
      </ul>
    </div>
  `;
},
data() {
  return {
    items: ['one', 'two', 'three'],
  }
},
...
```

**Comment by russiann** [comment]
I found a workaround to 'fix' this sad thing about JSX:

```js

render() {
  const events = {
    '@click': 'method'
    '@foo:bar:buzz': 'method'
  };

  return (
    <div class="page">
      <button {...events}>Click </button>
    </div>
  )

}

```

So, I can do something like this:

```js
const attr = (name, value) => {
  return {[name]: value}
}

...

<button {...attr('@click', 'method')} > Click </button>
```
Not ideal but works :)

**Comment by russiann** [comment]
I know about es6 template but I want to use vhtml.js with JSX because it allows me to create Pure Functional Components like this:

```js
let items = ['one', 'two'];

const Item = ({ item, index, children }) => (
  <li id={index}>
    <h4>{item}</h4>
    {children}
  </li>
);

const Element = (
  <div class="foo">
    <h1>Hi!</h1>
    <ul>
      { items.map( (item, index) => (
        <Item {...{ item, index }}>
          This is item {item}!
        </Item>
      )) }
    </ul>
  </div>
);
```

and this + f7 = awesome :D

**Comment by russiann** [comment]
I changed the issue name to make the search easier.
Maybe other developers would like to use JSX as well and now they can easily found this issue.
